File: src/legacy/unit_item_equivalent_4e7de0.c
#include "GAME3_3.h"

enum {
	item_equivalent_info_book_class_4e7de0 = 0x00000100,
	item_equivalent_modifier_class_mask_4e7de0 = 0x13001000,
};

//----- (004E7DE0) --------------------------------------------------------
int32_t sub_4E7DE0(const nox_object_t* candidate, const nox_object_t* item) {
	if (!candidate || !item) {
		return 0;
	}
	if (candidate->typ_ind != item->typ_ind) {
		return 0;
	}

	const uint32_t object_class = candidate->obj_class;
	if ((object_class & (uint32_t)item_equivalent_modifier_class_mask_4e7de0) != 0) {
		const nox_modifier_attrs_t* candidate_attrs = candidate->init_data;
		const nox_modifier_attrs_t* item_attrs = item->init_data;
		for (int i = 0; i < 4; ++i) {
			if (candidate_attrs->modifiers[i] != item_attrs->modifiers[i]) {
				return 0;
			}
		}
	}

	if ((object_class & (uint32_t)item_equivalent_info_book_class_4e7de0) == 0) {
		return 1;
	}

	const uint32_t object_subclass = candidate->obj_subclass;
	if ((object_subclass & UINT32_C(1)) != 0) {
		const uint8_t* candidate_use = candidate->use_data;
		const uint8_t* item_use = item->use_data;
		return candidate_use[0] == item_use[0];
	}
	if ((object_subclass & UINT32_C(2)) != 0) {
		const uint8_t* item_use = item->use_data;
		const uint8_t* candidate_use = candidate->use_data;
		for (;;) {
			const uint8_t candidate_byte = *candidate_use++;
			const uint8_t item_byte = *item_use++;
			if (candidate_byte != item_byte) {
				return 0;
			}
			if (candidate_byte == 0) {
				return 1;
			}
		}
	}

	const uint8_t* candidate_use = candidate->use_data;
	const uint8_t* item_use = item->use_data;
	return candidate_use[0] == item_use[0];
}

File: src/legacy/unit_item_equivalent_4e7de0.go
package legacy

const (
	itemEquivalentInfoBookClass4E7DE0      = uint32(0x00000100)
	itemEquivalentModifierClass4E7DE0      = uint32(0x13001000)
	itemEquivalentSpellBookSubclass4E7DE0  = uint32(0x00000001)
	itemEquivalentFieldGuideSubclass4E7DE0 = uint32(0x00000002)
)

type itemEquivalentHooks4E7DE0[O, A, U, M comparable] struct {
	loadType     func(O) uint16
	loadClass    func(O) uint32
	loadInitData func(O) A
	loadModifier func(A, int) M
	loadSubclass func(O) uint32
	loadUseData  func(O) U
	loadUseByte  func(U, int) byte
}

// itemEquivalent4E7DE0 preserves GAME.EXE 004E7DE0. The candidate alone
// selects the modifier and InfoBook comparison modes. Modifier-bearing items
// compare four native pointer identities. SpellBooks compare one use-data
// byte, FieldGuides compare a NUL-terminated byte string, and other InfoBooks
// compare one byte. Type, field, and byte loads retain the original
// short-circuit order.
func itemEquivalent4E7DE0[O, A, U, M comparable](
	candidate O,
	item O,
	hooks itemEquivalentHooks4E7DE0[O, A, U, M],
) bool {
	var zero O
	if candidate == zero || item == zero {
		return false
	}
	if hooks.loadType(candidate) != hooks.loadType(item) {
		return false
	}

	objectClass := hooks.loadClass(candidate)
	if objectClass&itemEquivalentModifierClass4E7DE0 != 0 {
		candidateAttrs := hooks.loadInitData(candidate)
		itemAttrs := hooks.loadInitData(item)
		for i := range 4 {
			candidateModifier := hooks.loadModifier(candidateAttrs, i)
			itemModifier := hooks.loadModifier(itemAttrs, i)
			if candidateModifier != itemModifier {
				return false
			}
		}
	}

	if objectClass&itemEquivalentInfoBookClass4E7DE0 == 0 {
		return true
	}

	objectSubclass := hooks.loadSubclass(candidate)
	if objectSubclass&itemEquivalentSpellBookSubclass4E7DE0 != 0 {
		candidateUse := hooks.loadUseData(candidate)
		itemUse := hooks.loadUseData(item)
		candidateByte := hooks.loadUseByte(candidateUse, 0)
		itemByte := hooks.loadUseByte(itemUse, 0)
		return candidateByte == itemByte
	}
	if objectSubclass&itemEquivalentFieldGuideSubclass4E7DE0 != 0 {
		itemUse := hooks.loadUseData(item)
		candidateUse := hooks.loadUseData(candidate)
		for i := 0; ; i++ {
			candidateByte := hooks.loadUseByte(candidateUse, i)
			itemByte := hooks.loadUseByte(itemUse, i)
			if candidateByte != itemByte {
				return false
			}
			if candidateByte == 0 {
				return true
			}
		}
	}

	candidateUse := hooks.loadUseData(candidate)
	itemUse := hooks.loadUseData(item)
	candidateByte := hooks.loadUseByte(candidateUse, 0)
	itemByte := hooks.loadUseByte(itemUse, 0)
	return candidateByte == itemByte
}
